### DISCUSSION TITLE
When to use SWR or redux async actions (eg Thunk)

### DISCUSSION THREAD

**Discussion by getorca** [vercel/next.js#14601] [Help]
I'm struggling to decide on when to use SWR or async actions in redux. 

For example my observeUser flow for anonymous users:

**Case 1: redux async actions and try to handle side effects in actions**
```
// in my layout wrapper
    useEffect(() => {        
        observeUser()      // dispacther to async action
    }, [])

   useEffect(() => {        
        // handle redirects / modal or could be put in async action with https://github.com/danielr18/connected-next-router
    }, [user])
```
```
// in my user actions
export function observeUser(payload) {   
    return dispatch => {
        fetch('/api/auth/user')
        .then(res => res.json())
        .then(res => {
            dispatch(updateUser(res))
            return res;
        })
    }
}

export function updateUser(payload) {
    return { type: userActionTypes.UPDATE_USER, payload }
}

```

**Case 2: useUser example with SWR, and useEffect updates global state**
see: https://github.com/vercel/next.js/tree/canary/examples/with-iron-session

```
// in my layout component
  const { user } = useUser({})   
  useEffect(() => {
        if(user) {
            updateUser(user) // dispatch to update the redux store
        }      
    }, [user])
```

**🤔 Thoughts...**
Clearly using SWR makes the most sense when I don't need to update global state. But in cases where I need global state I can see pros and cons of both cases.
- using SWR can leave the redux pure to only update the store.
- using redux async actions can move a lot of code out of UI components. Additional middleware can enhance this but lead to more confusing actions, eg: for the same global state some pages/components might need to push router changes while others don't.
- a combination of both where appropriate, probably the most likely to lead to bugs out of confusion, and probably bad for on-boarding new devs, eg why did they do it this way here and not there.
- SWR when global state's not updated, async actions when global state needs to mutate?

Thoughts? Or how do you handle it in your application?

**Comment by vvo** [comment]
Hey there, I was just searching for people mentioning my project. Quick thoughts: I have been building complete React applications only using SWR, not even redux. So if you're just starting your project I would advise you give this idea a try: use only SWR for data fetching.

You can have a useUser hook that is accessible from anywhere. You can mutate your user and have it updated in every component. Let me know what you think

**Comment by saisandeepvaddi** [comment]
Hey,

I no longer use Redux where I need to get data from server even if it is global state.

I do something like following.
```jsx
//global-data-store.jsx
const GlobalDataContext = React.createContext({
  users: []
});

export function GlobalDataProdiver(props) {
 const {data: users} = useSWR('/api/users', fetcher);
 const {data: items} = useSWR('/api/items, fetcher);

 const value = React.useMemo(() => ({users, items}), [users, items]);
 return <GlobalDataContext.Provider value={value} {...props} />
}

export function useGlobalData() {
 const context = React.useContext(GlobalDataContext);
 if(!context) throw new Error("Need to wrap GlobalDataProdiver");
 return context;
}

```
```jsx
// _app.jsx or any top-level component

// Wrap component with <GlobalDataProvider>

 function MyTopLevelComponent() {
  ----
  return <GlobalDataProvider>
//// your chlildren
</GlobalDataProvider>
}
```
Now wherever you need global data, just use `useGlobalData` hook.

``` jsx
// MyUserChildComponent.jsx

function MyUserChildComponent() {
  const {users} = useGlobalData();
  return <div>
      {users.map(user => <div key={user.id}>{user.name}</div>}
     </div>
}

```
``` jsx
// MyItemChildComponent.jsx

function MyItemChildComponent() {
  const {items} = useGlobalData();
  return <div>
      {item.map(user => <div key={item.id}>{item.name}</div>}
     </div>
}
```

For me, this is much simpler. The best thing is whenever the data updates or anything that mutates `/api/users` and `/api/items`, you don't have to push to redux store. It automatically gets updated to all components where the `useGlobalData` is used. You don't need observers and manually update the redux store.

You can also use `useSWR`'s global `mutate('/api/users')` anywhere in the app to force update & revalidate state or you can even pass bound `mutate` functions right from the `GlobalDataProvider` like 

```jsx
//global-data-store.jsx

export function GlobalDataProdiver(props) {
 const {data: users, mutate: updateUsers} = useSWR('/api/users', fetcher);
----

 const value = React.useMemo(() => ({users, items, updateUsers}), [users, items, updateUsers]);
 return <GlobalDataContext.Provider value={value} {...props} />
}

```

**Reply by windmaomao** [reply]
kool, just a quick question, what's the reason to use `useContext` to provide the `GlobalDataProvider`, because `swr` is already global, and simply just use `const { data } = useSWR or useGlobalData`.

is this extra layer? The only benefits i can see is that you don't want other developer to abuse `useGlobalData` whenever they can.

**Reply by saisandeepvaddi** [reply]
You don't need actually. Like you said it's an extra layer. I used it because I was using ContextProvider for other things. It's may be not needed. Check @TommySorensen 's answer which is simpler.

**Reply by theonlydaleking** [reply]
Can you update the first code block to fix that missing quote mark? Makes reading easier when the syntax highlighting isn't broken

**Reply by Iyosayi1** [reply]
What about paginated data ?

**Comment by TommySorensen** [comment]
There is no need to create a context provider for the SWR data. You can get that out of the box by putting the SWR in its own hook with something like this:

```js
// useSomeData.js
const useSomeData = () => {
  const { data, error, mutate } = useSWR('/api/something')

  return {
    data,
    isLoading: !error && !data,
    isError: error,
    mutate
  }
}

export default useSomeData
```

```js
// In some component
const { data, isLoading, isError } = useSomeData()
```
```js
// In some other component
const { data, isLoading, isError } = useSomeData()
```

The data is shared between the 2 components. Simple as that 🙂

**Reply by markmendozadev** [reply]
very useful thanks.

**Comment by windmaomao** [comment]
I searched the internet for `useSWR`, `redux` or `useContext`. I especially find some similarity between `SWR` and `Context` with usage of `setState`.

#### onUpdate

Each `component` registers some sort of `onUpdate` to the pool, although `SWR` seems to be registered globally while `Context` is relatively local to the ancestor. But more or less if the `value` changes, it'll notify all registered listener by calling `setState` which in turn triggers the element update. Hopefully the batched update would make sure we end up with only one big render.

#### initial update

Upon initial update, `useContext` communicates to the ancestor directly to get the `value`, while `useSWR` look into the cache before invalidating it.  Both can get the `value` quickly if provided already. 


#### NOTES

I'm no expert on either approach. Just from the surface, it seems `useSWR` could be flexible enough to replace `redux`. From the past, seems to me `redux` isn't very targeted and very verbose to use, but `useSWR` seems rely on the `key` and is only one line to use.  The only minor drawback i can see using `useSWR` is that the pool is globally defined, but that's same as `redux`.

**Reply by zhenghao1** [reply]
I think definitely there's an apparent convenience gained with the use of useSWR.  It definitely makes calling external APIs much simpler without the need for context or prop drilling.  

However, I wouldn't throw away redux just yet.  I think redux has it's purpose and value.  For small applications I would think that redux is overkill and that SWR would actually fit it quite well.   For large applications, I think it would be a management nightmare trying to figure out where your component got its state from.  In this case, having a centralized global state store like redux makes things much more manageable.  

Plus these days with Redux Toolkit, it's not THAT hard to use redux anymore.  

Just need to assess the complexity of the project you are building and make the best educated choice.

**Comment by codthing** [comment]
Geez, what are you guys talking about, did I miss something? SWR and Redux are not the same thing. swr is only for data requests in my opinion. Whereas Redux is used to manage all state changes. I hope I'm understanding this correctly.

**Comment by phoenixcoded20** [comment]
Well, there is no where mention that SWR cannot be used as state management as well. You can use key that can be also non api thing and use it as state management. See entire thread here: https://github.com/vercel/swr/discussions/587

**Reply by caglartufan** [reply]
This might be correct but SWR allows you to revalidate data by calling the mutate function which fetches the data again for you, automatically or upon your need. If you were to use SWR for local states, non API data then you would be using wrong tool. For local state, Context API or Redux or other similar solutions are preferred.
